crates/environ/src/fact/trampoline.rs

```diff
@@ -256,8 +402,282 @@ struct Memory<'a> {
+
+        self.finish()
+    }
+
+    fn compile_sync_to_async_adapter(
+        mut self,
+        adapter: &AdapterData,
+        start: FunctionId,
+        return_: FunctionId,
+        param_count: i32,
+        param_globals: Option<&[u32]>,
+        result_globals: Option<&[u32]>,
+    ) {
+        let enter = self.module.import_async_enter_call();
+        let exit = self
+            .module
+            .import_async_exit_call(adapter.lift.options.callback, None);
+
+        self.flush_code();
+        self.module.funcs[self.result]
+            .body
+            .push(Body::RefFunc(start));
+        self.module.funcs[self.result]
+            .body
+            .push(Body::RefFunc(return_));
+        self.instruction(I32Const(
+            i32::try_from(adapter.lower.instance.as_u32()).unwrap(),
+        ));
+        self.instruction(I32Const(
+            i32::try_from(self.types[adapter.lift.ty].params.as_u32()).unwrap(),
+        ));
+
+        let results_local = if let Some(globals) = param_globals {
+            for (local, global) in globals.iter().enumerate() {
+                self.instruction(LocalGet(u32::try_from(local).unwrap()));
+                self.flush_code();
+                self.module.funcs[self.result]
+                    .body
+                    .push(Body::GlobalSet(*global));
+            }
+            self.instruction(I32Const(0)); // dummy params pointer
+            u32::try_from(globals.len()).unwrap()
+        } else {
+            self.instruction(LocalGet(0));
+            1
+        };
+
+        if result_globals.is_some() {
+            self.instruction(I32Const(0)); // dummy results pointer
+        } else {
+            self.instruction(LocalGet(results_local));
+        }
+
+        self.instruction(Call(enter.as_u32()));
+
+        // TODO: As an optimization, consider checking the backpressure flag on the callee instance and, if it's
+        // unset _and_ the callee uses a callback, translate the params and call the callee function directly here
+        // (and make sure `exit` knows _not_ to call it in that case).
+
+        self.module.exports.push((
+            adapter.callee.as_u32(),
+            format!("[adapter-callee]{}", adapter.name),
+        ));
+        self.instruction(I32Const(
+            i32::try_from(adapter.lower.instance.as_u32()).unwrap(),
+        ));
+        self.instruction(RefFunc(adapter.callee.as_u32()));
+        self.instruction(I32Const(
+            i32::try_from(adapter.lift.instance.as_u32()).unwrap(),
+        ));
+        self.instruction(I32Const(param_count));
+        self.instruction(I32Const(1)); // leave room for the guest context result
+        self.instruction(I32Const(EXIT_FLAG_ASYNC_CALLEE));
+        self.instruction(Call(exit.as_u32()));
+        self.instruction(Drop);
+
+        if let Some(globals) = result_globals {
+            for global in globals {
+                self.global_set(*global);
+            }
+        }
```

Comment:
I fear I'm personally still pretty lost here. In a sync->async call what I was expecting is that the results would be stored in globals and async-start would read from those globals when called by somone (the host? unsure). The results would then be stored-to in async-return. Here I was expecting something to block (as the caller is sync) until the nested async call was finished and then the globals would be read.

That doesn't appear to be what's happening though? I commented above that's it's not clear what async-start/async-return are doing but the setting of globals here rather than reading is what really threw me astray... Mind leaving more comments as to what the host intrinsics are doing and the various protocol that this is adhering to?

Reply:
I've refactored this to avoid using globals, and have added explanatory comments to the new code.